### PR TITLE
Buffered Transactional Editing (gui/settings part)

### PR DESCRIPTION
**Pull request by domi4484**
## Buffered Transactional Editing

Implementation of [QGIS-Enhancement-Proposals #203](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/203)
This pr includes only the gui/settings part and is followed by #46750

image

### CODE REVIEW

**Comment by domi4484** [comment]
@3nids I have started with the buffered transactions implementation, just the gui settings part for the moment.
Few questions/remarks:
- The property `autotransaction` was renamed to `transaction-mode`. Opening old project it will check status of the old property if the new one does not exist
- There are three different transaction modes:
  - **None** Edits are buffered locally and sent to the provider when toggling layer editing mode (default as it is now).
  - **AutomaticGroups** Automatic transactional editing means that on supported datasources (postgres databases) the edit state of all tables that originate from the same database are synchronized and executed in a server side transaction (as it is now).
  - **BufferedGroups** Buffered transactional editing means that all editable layers in the buffered transaction group are toggled synchronously and all edits are saved in a local edit buffer. Saving changes is executed within a single transaction on all layers (per provider) (new mode).
- I think there was a way to save enum properties as string but I can't remenber how right now?

File: src/core/project/qgsproject.cpp [hunk 1]
```diff
@@ -2293,8 +2302,8 @@ bool QgsProject::writeProjectFile( const QString &filename )
   QDomElement titleNode = doc->createElement( QStringLiteral( "title" ) );
   qgisNode.appendChild( titleNode );
 
-  QDomElement transactionNode = doc->createElement( QStringLiteral( "autotransaction" ) );
-  transactionNode.setAttribute( QStringLiteral( "active" ), mAutoTransaction ? 1 : 0 );
+  QDomElement transactionNode = doc->createElement( QStringLiteral( "transaction-mode" ) );
+  transactionNode.setAttribute( QStringLiteral( "active" ), QString::number( static_cast<int>( mTransactionMode ) ) );
```
Comment (3nids) [inline src/core/project/qgsproject.cpp:2306]:
```suggestion
  QDomElement transactionNode = doc->createElement( QStringLiteral( "transaction" ) );
  transactionNode.setAttribute( QStringLiteral( "mode" ), qgsEnumValueToKey( mTransactionMode ) );
```

File: src/core/project/qgsproject.cpp [hunk 2]
```diff
@@ -1575,11 +1575,20 @@ bool QgsProject::readProjectFile( const QString &filename, QgsProject::ReadFlags
   }
   emit metadataChanged();
 
-  element = doc->documentElement().firstChildElement( QStringLiteral( "autotransaction" ) );
-  if ( ! element.isNull() )
+  // Transaction mode
+  element = doc->documentElement().firstChildElement( QStringLiteral( "transaction-mode" ) );
```
Comment (3nids) [inline src/core/project/qgsproject.cpp:1579]:
```suggestion
  element = doc->documentElement().firstChildElement( QStringLiteral( "transaction" ) );
```

File: src/core/project/qgsproject.cpp [hunk 3]
```diff
@@ -1575,11 +1575,20 @@ bool QgsProject::readProjectFile( const QString &filename, QgsProject::ReadFlags
...
+  if ( !element.isNull() )
   {
-    if ( element.attribute( QStringLiteral( "active" ), QStringLiteral( "0" ) ).toInt() == 1 )
-      mAutoTransaction = true;
+    mTransactionMode = static_cast<Qgis::TransactionMode>( element.attribute( QStringLiteral( "active" ), QStringLiteral( "0" ) ).toInt() );
```
Comment (3nids) [inline src/core/project/qgsproject.cpp:1582]:
```suggestion
    mTransactionMode = qgsEnumKeyToValue<Qgis::TransactionMode>( element.attribute( QStringLiteral( "mode" ), QSgring() ), Qgis::TransactionMode::None );
```

File: src/core/qgis.h [hunk 4]
```diff
@@ -1248,6 +1248,26 @@ class CORE_EXPORT Qgis
...
     Q_ENUM( AngularDirection )
 
+    /**
+     * Transaction mode.
+     *
+     * \since QGIS 3.24
+     */
+    enum class TransactionMode : int
+    {
+      None = 0, /*!< Edits are buffered locally and sent to the provider when toggling
```
Comment (3nids) [inline src/core/qgis.h:1258]: would `Disabled` be more appropriate?

**Review by 3nids — COMMENTED** [review]
Looks good!

File: src/app/qgsprojectproperties.cpp [hunk 5]
```diff
@@ -168,6 +168,13 @@ QgsProjectProperties::QgsProjectProperties( QgsMapCanvas *mapCanvas, QWidget *pa
   mAreaUnitsCombo->addItem( tr( "Square Degrees" ), QgsUnitTypes::AreaSquareDegrees );
   mAreaUnitsCombo->addItem( tr( "Map Units" ), QgsUnitTypes::AreaUnknownUnit );
 
+  mTransactionModeComboBox->addItem( tr( "Local edit buffer" ), static_cast< int >( Qgis::TransactionMode::None ) );
```
Comment (nyalldawson) [inline src/app/qgsprojectproperties.cpp:171]: Combo text should be in Title Case

File: src/core/project/qgsproject.cpp [hunk 6]
```diff
@@ -3272,20 +3281,47 @@ QStringList QgsProject::nonIdentifiableLayers() const
...
+}
+
+void QgsProject::setTransactionMode( Qgis::TransactionMode transactionMode )
+{
+  if ( transactionMode == mTransactionMode )
+    return;
+
+  mTransactionMode = transactionMode;
+
+  if ( mTransactionMode == Qgis::TransactionMode::AutomaticGroups )
```
Comment (nyalldawson) [inline src/core/project/qgsproject.cpp:3321]: This should be a switch statement (with no default case, so warnings are raised if new values are ever added to the enum)

File: src/core/project/qgsproject.h [hunk 7]
```diff
@@ -885,8 +886,27 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
...
+     *
+     * \see Qgis::TransactionMode
+     * \since QGIS 3.24
+     */
+    Qgis::TransactionMode transactionMode() const;
+
+    /**
+     * Set transaction mode
+     *
+     * Make sure that this is only called when all layers are not in edit mode.
```
Comment (nyalldawson) [inline src/core/project/qgsproject.h:904]: Should an exception be raised if this condition is not met?

Comment (nyalldawson) [inline src/core/project/qgsproject.h:904]:
```suggestion
     * \warning Make sure that this is only called when all layers are not in edit mode.
```

File: src/core/qgis.h (hunk 4 above)
Comment (nyalldawson) [inline src/core/qgis.h:1258]: Sipify doesn't support multiline documentation like this -- you have to make it all one giant single line.

Comment (m-kuhn) [reply src/core/qgis.h:1258]: can it be added before the enum value as block comment=

Comment (nyalldawson) [reply src/core/qgis.h:1258]: Not without sipify modifications 😱😜

File: src/core/project/qgsproject.cpp (hunk 3 above)
Comment (domi4484) [reply src/core/project/qgsproject.cpp:1582]: Thanks, no need to explicitly specify the template `<Qgis::TransactionMode>` as the compiler can deduce it from the default value

File: src/core/project/qgsproject.h (hunk 7 above)
Comment (domi4484) [reply src/core/project/qgsproject.h:904]: Probably yes, but looking at the code I don't understand exactly why. Probably unsaved modification would be lost when the transaction groups are cleared.

Comment (domi4484) [reply src/core/project/qgsproject.h]: I have made the function returning false if at leas one layer is in edit mode.

**Comment by domi4484** [comment]
All the commits of this pr are also included in #46750, so this one can stay closed.

**Pull request closed without merging**